parquet/src/test/java/org/apache/iceberg/parquet/TestVariantReadsFromFile.java

```diff
@@ -0,0 +1,198 @@
+              String parquetFile = JsonUtil.getStringOrNull("parquet_file", caseNode);
+              String variantFile = JsonUtil.getStringOrNull("variant_file", caseNode);
+              return Arguments.of(caseNumber, testName, variant, parquetFile, variantFile);
+            });
+  }
+
+  private static Stream<Arguments> multiVariantCases() throws IOException {
+    return cases()
+        .filter(caseNode -> caseNode.has("variant_files") || !caseNode.has("parquet_file"))
+        .map(
+            caseNode -> {
+              int caseNumber = JsonUtil.getInt("case_number", caseNode);
+              String testName = JsonUtil.getStringOrNull("test", caseNode);
+              String parquetFile = JsonUtil.getStringOrNull("parquet_file", caseNode);
+              List<String> variantFiles =
+                  caseNode.has("variant_files")
+                      ? Lists.newArrayList(
+                          Iterables.transform(
+                              caseNode.get("variant_files"),
+                              node -> node == null || node.isNull() ? null : node.asText()))
+                      : null;
+              String variants = JsonUtil.getStringOrNull("variants", caseNode);
+              return Arguments.of(caseNumber, testName, variants, parquetFile, variantFiles);
+            });
+  }
+
+  @ParameterizedTest
+  @MethodSource("errorCases")
+  public void testError(int caseNumber, String testName, String parquetFile, String errorMessage) {
+    if (parquetFile == null) {
+      return;
+    }
+
+    Assertions.assertThatThrownBy(() -> readParquet(parquetFile))
+        .as("Test case %s: %s", caseNumber, testName)
+        .hasMessageContaining(errorMessage);
+  }
+
+  @ParameterizedTest
+  @MethodSource("singleVariantCases")
+  public void testSingleVariant(
+      int caseNumber, String testName, String variant, String parquetFile, String variantFile)
+      throws IOException {
+    if (parquetFile == null) {
+      return;
+    }
+
+    Variant expected = readVariant(variantFile);
+
+    Record record = readParquetRecord(parquetFile);
+    Assertions.assertThat(record.getField("var")).isInstanceOf(Variant.class);
+    Variant actual = (Variant) record.getField("var");
+    VariantTestUtil.assertEqual(expected.metadata(), actual.metadata());
+    VariantTestUtil.assertEqual(expected.value(), actual.value());
+  }
+
+  @ParameterizedTest
+  @MethodSource("multiVariantCases")
+  public void testMultiVariant(
+      int caseNumber,
+      String testName,
+      String variants,
+      String parquetFile,
+      List<String> variantFiles)
+      throws IOException {
+    if (parquetFile == null) {
+      return;
+    }
+
+    List<Record> records = readParquet(parquetFile);
+
+    for (int i = 0; i < records.size(); i += 1) {
+      Record record = records.get(i);
+      String variantFile = variantFiles.get(i);
+
+      if (variantFile != null) {
```

Comment:
Nit: for readability, consider flipping the order:

if (variantFile == null) {
   // null logic here
} else {
   // non-null logic here.
}